### PR TITLE
[macOS, multiwindow] Compositor gets FlutterView lazily

### PR DESCRIPTION
**Pull request by dkwingsmt**
This PR changes `macos/FlutterCompositor` so that instead of getting `FlutterViewController*` at construction, it gets a callback that provides the `FlutterView*` lazily.

Part of [the multiwindow project](https://github.com/flutter/flutter/issues/30701) ([design doc](https://docs.google.com/document/d/1U_ciMTHJgDrnw8kIo9sGKhiKFOycWRmz1yVVI_osCXg/edit)): This is a necessary step for multi-window because `FlutterCompositor` is [constructed at engine initialization](https://github.com/flutter/engine/blob/main/shell/platform/darwin/macos/framework/Source/FlutterEngine.mm#L355), while there can be multiple views or view controllers added after initialization.

Existing tests are adapted. No tests are added.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on
writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: shell/platform/darwin/macos/framework/Source/FlutterEngine.mm [hunk 1]
```diff
@@ -428,16 +428,22 @@ - (FlutterCompositor*)createFlutterCompositor {
...
-  if (!_viewController) {
+  FlutterViewController* viewController = _viewController;
+  if (!viewController) {
     return nil;
   }
 
   __weak FlutterEngine* weakSelf = self;
 
+  flutter::FlutterCompositor::GetViewCallback get_view_callback =
+      [&viewController](uint64_t view_id) {
```
Comment (stuartmorgan) [inline shell/platform/darwin/macos/framework/Source/FlutterEngine.mm:439]: You're capturing a fixed snapshot of `_viewController` here, and I don't see anything in `setViewController` that would update this, so it looks like you're going to end up with stale references any time the view controller for the engine changes.

File: shell/platform/darwin/macos/framework/Source/FlutterEngine.mm [hunk 2]
```diff
@@ -428,16 +428,22 @@ - (FlutterCompositor*)createFlutterCompositor {
...
   // https://github.com/flutter/flutter/issues/71606
-  if (!_viewController) {
+  FlutterViewController* viewController = _viewController;
+  if (!viewController) {
     return nil;
   }
 
   __weak FlutterEngine* weakSelf = self;
 
+  flutter::FlutterCompositor::GetViewCallback get_view_callback =
```
Comment (stuartmorgan) [inline shell/platform/darwin/macos/framework/Source/FlutterEngine.mm:438]: Nit: ObjC++ Google style is to match the naming style to the enclosing method, not the variable type. So a C++ type in an ObjC method is named using objCNamingStyle not cpp_naming_style.

**Comment by zanderso** [comment]
From PR triage: @dkwingsmt is this ready for another review pass from @stuartmorgan?

**Comment by dkwingsmt** [comment]
@zanderso @.stuartmorgan said he did not know enough to review this PR. Can you help me find someone to review it?

File: shell/platform/darwin/macos/framework/Source/FlutterCompositor.h [hunk 3]
```diff
@@ -19,7 +19,9 @@ namespace flutter {
 // Platform views are not yet supported.
 class FlutterCompositor {
  public:
-  explicit FlutterCompositor(FlutterViewController* view_controller);
+  using GetViewCallback = std::function<FlutterView*(uint64_t view_id)>;
```
Comment (iskakaushik) [inline shell/platform/darwin/macos/framework/Source/FlutterCompositor.h:22]: I think `ViewProvider` is a better name for this than a view callback.

**Comment by iskakaushik** [comment]
I'm trying to understand how this would get used in the multi-window context, can you provide a brief sketch of how you would end up wiring the get view callback when there are multiple views at play?

**Comment by dkwingsmt** [comment]
Sure!

First of all, here is a design doc for the mac engine multi-window project. It's still in progress, but contains sufficient information till this PR. https://docs.google.com/document/d/1U_ciMTHJgDrnw8kIo9sGKhiKFOycWRmz1yVVI_osCXg/edit

Second, in a multi-window world,
- A view controller corresponds to a view
- The Compositor class is a singleton. 
(Both for reasons written in the design doc in the table.)
Since views can be added after the app has been started, the compositor must be able to handle views that are added after its contruction, instead of just the one from the constructor.

How will the compositor get view controllers in the future? The compositor is constructed by the FlutterEngine, and receives the ViewProvider callback from the FlutterEngine. The engine currently stores a single viewController, but in the future it will store a map of view controllers, mapped from surface IDs. The ViewProvider will use the argument Id to find the correct view.

The ID is also described in the design doc. (I should probably have provided the doc in the first place, but the doc was still being written when this PR was drafted.)

File: shell/platform/darwin/macos/framework/Source/FlutterEngine.mm [hunk 4]
```diff
@@ -434,14 +434,18 @@ - (FlutterCompositor*)createFlutterCompositor {
 
   __weak FlutterEngine* weakSelf = self;
 
+  flutter::FlutterCompositor::ViewProvider getViewCallback = [weakSelf](uint64_t view_id) {
+    return weakSelf.viewController == nullptr ? nullptr : weakSelf.viewController.flutterView;
```
[the remark this answers is not in the document]
Comment (knopp) [reply shell/platform/darwin/macos/framework/Source/FlutterEngine.mm:438]: Is that necessary? In objc sending messages to nil object returns nil. Wouldn't just `return weakSelf.viewController.flutterView` be enough?

[the remark this answers is not in the document]
Comment (chinmaygarde) [reply shell/platform/darwin/macos/framework/Source/FlutterEngine.mm:438]: @knopp is right. @iskakaushik is perhaps thinking of the fml::WeakPtrs?

File: shell/platform/darwin/macos/framework/Source/FlutterCompositor.h [hunk 5]
```diff
@@ -62,7 +64,7 @@ class FlutterCompositor {
   typedef enum { kStarted, kPresenting, kEnded } FrameStatus;
 
  protected:
-  __weak const FlutterViewController* view_controller_;
+  FlutterView* GetView(uint64_t view_id);
```
Comment (stuartmorgan) [inline shell/platform/darwin/macos/framework/Source/FlutterCompositor.h:67]: Nit: Per style guide, this needs a declaration comment explaining it.

File: shell/platform/darwin/macos/framework/Source/FlutterCompositor.h [hunk 6]
```diff
@@ -76,15 +78,18 @@ class FlutterCompositor {
...
   void InsertCALayerForIOSurface(
+      FlutterView* view,
       const IOSurfaceRef& io_surface,
       CATransform3D transform = CATransform3DIdentity);
 
  private:
   // A list of the active CALayer objects for the frame that need to be removed.
   std::list<CALayer*> active_ca_layers_;
 
+  ViewProvider get_view_callback_;
```
Comment (stuartmorgan) [inline shell/platform/darwin/macos/framework/Source/FlutterCompositor.h:91]: Ideally this should have a declaration comment explain what its role is as well.

File: shell/platform/darwin/macos/framework/Source/FlutterCompositor.h [hunk 7]
```diff
@@ -76,15 +78,18 @@ class FlutterCompositor {
   bool EndFrame(bool has_flutter_content);
 
   // Creates a CALayer object which is backed by the supplied IOSurface, and
-  // adds it to the root CALayer for this FlutterViewController's view.
+  // adds it to the root CALayer for the given view.
   void InsertCALayerForIOSurface(
+      FlutterView* view,
```
Comment (stuartmorgan) [inline shell/platform/darwin/macos/framework/Source/FlutterCompositor.h:83]: What's the purpose of this change? If the compositor is inherently bound to a view provider at creation time, why do we want public API that uses an arbitrary view rather than the view provided by its provider? It seems like this makes it much easier to accidentally use the wrong view object here. Sholudn't it take a view ID instead?

Comment (dkwingsmt) [reply shell/platform/darwin/macos/framework/Source/FlutterCompositor.h:82]: The very change of this PR is that the compositor is no longer bound to a view provider at creation. The compositor (which needs to be a singleton) needs to be able to handle multiple views that are only available after the app starts.

Comment (dkwingsmt) [reply shell/platform/darwin/macos/framework/Source/FlutterCompositor.h:82]: Moreover, `InsertCALayerForIOSurface` knows nothing about where the view comes from. `FlutterCompositor` only knows that there's a new CALayer that needs to be removed later. The only result that an evil subclass calls this method with a strange view is that the CALayer will be removed.

**Comment by skia-gold** [comment]
Gold has detected about 2 new digest(s) on patchset 17.
View them at https://flutter-engine-gold.skia.org/cl/github/36392

File: shell/platform/darwin/macos/framework/Source/FlutterCompositor.h (hunk 7 above)
Comment (stuartmorgan) [reply shell/platform/darwin/macos/framework/Source/FlutterCompositor.h:90]: Sorry, I missed that this was protected; I thought it was fully public API.

File: shell/platform/darwin/macos/framework/Source/FlutterGLCompositor.mm [hunk 8]
```diff
@@ -19,13 +19,16 @@
...
                                          NSOpenGLContext* opengl_context)
-    : FlutterCompositor(view_controller), open_gl_context_(opengl_context) {}
+    : FlutterCompositor(view_provider), open_gl_context_(opengl_context) {}
 
 bool FlutterGLCompositor::CreateBackingStore(const FlutterBackingStoreConfig* config,
                                              FlutterBackingStore* backing_store_out) {
-  if (!view_controller_) {
+  // Always gets the first view, #0. After Flutter supports multi-view, it
+  // should get the view ID from somewhere.
+  FlutterView* view = GetView(0);
```
Comment (stuartmorgan) [inline shell/platform/darwin/macos/framework/Source/FlutterGLCompositor.mm:30]: Shouldn't this be using the constant?

File: shell/platform/darwin/macos/framework/Source/FlutterViewProvider.h [hunk 9]
```diff
@@ -0,0 +1,29 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#import "flutter/shell/platform/darwin/macos/framework/Source/FlutterView.h"
+
+@class FlutterEngine;
+
+/**
+ * @brief A facade over FlutterEngine that allows FlutterEngine's children
```
Comment (stuartmorgan) [inline shell/platform/darwin/macos/framework/Source/FlutterViewProvider.h:10]: Is there a reason this is so tightly bound to `FlutterEngine`? Usually something like this would be a `@protocol`, and concrete implementations (like one bound to an engine object) would be created--potentially file-scoped--by the thing providing it.

So I would expect `@protocol FlutterViewProvider : NSObject`, and for there to be a private class called something like `EngineViewProvider` that implements this weak referencing delegation, which is an implementation detail.

File: shell/platform/darwin/macos/framework/Source/FlutterGLCompositor.mm (hunk 8 above)
Comment (iskakaushik) [reply shell/platform/darwin/macos/framework/Source/FlutterGLCompositor.mm:30]: @dkwingsmt I think you meant to use: kFlutterDefaultViewId here

File: shell/platform/darwin/macos/framework/Source/FlutterGLCompositorUnittests.mm [hunk 10]
```diff
@@ -3,18 +3,26 @@
...
 
 namespace flutter::testing {
+namespace {
 
-TEST(FlutterGLCompositorTest, TestPresent) {
-  id mockViewController = CreateMockViewController();
+FlutterViewProvider* MockViewProvider() {
+  id viewMock = OCMClassMock([FlutterView class]);
+  FlutterViewProvider* viewProviderMock = OCMStrictClassMock([FlutterViewProvider class]);
+  OCMStub([viewProviderMock getView:0]).ignoringNonObjectArgs().andReturn(viewMock);
```
Comment (iskakaushik) [inline shell/platform/darwin/macos/framework/Source/FlutterGLCompositorUnittests.mm:18]: use `kFlutterDefaultViewId` here

File: shell/platform/darwin/macos/framework/Source/FlutterMetalCompositor.mm [hunk 11]
```diff
@@ -11,16 +11,20 @@
...
     id<MTLDevice> mtl_device)
-    : FlutterCompositor(view_controller),
+    : FlutterCompositor(view_provider),
       mtl_device_(mtl_device),
       platform_views_controller_(platform_views_controller) {}
 
 bool FlutterMetalCompositor::CreateBackingStore(const FlutterBackingStoreConfig* config,
                                                 FlutterBackingStore* backing_store_out) {
-  if (!view_controller_) {
+  // TODO(dkwingsmt): This class only supports single-view for now. As more
```
Comment (iskakaushik) [inline shell/platform/darwin/macos/framework/Source/FlutterMetalCompositor.mm:23]: Is there an issue for overall multi-window effort for macOS. It will be useful to link here.

File: shell/platform/darwin/macos/framework/Source/FlutterMetalCompositor.mm [hunk 12]
```diff
@@ -11,16 +11,20 @@
...
+  // classes are gradually converted to multi-view, it should get the view ID
+  // from somewhere.
+  FlutterView* view = GetView(kFlutterDefaultViewId);
+  if (!view) {
```
Comment (iskakaushik) [inline shell/platform/darwin/macos/framework/Source/FlutterMetalCompositor.mm:27]: When can this happen?

Comment (dkwingsmt) [reply shell/platform/darwin/macos/framework/Source/FlutterMetalCompositor.mm:27]: When the caller sends a view ID that the engine does not recognize. When the engine is run as a whole, this should not happen.

File: shell/platform/darwin/macos/framework/Source/FlutterMetalCompositor.mm (hunk 11 above)
Comment (dkwingsmt) [reply shell/platform/darwin/macos/framework/Source/FlutterMetalCompositor.mm:23]: We only have an overall multiwindow issue. Do you think it's worth opening one for the macOS multi-surface?

File: shell/platform/darwin/macos/framework/Source/FlutterViewProvider.h (hunk 9 above)
Comment (dkwingsmt) [reply shell/platform/darwin/macos/framework/Source/FlutterViewProvider.h]: Done.

**Review by iskakaushik — APPROVED** [review]
LGTM
